src/internal_modules/roc_core/string_buffer.cpp

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2020 Roc Streaming authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <roc_core/string_buffer.h>
+
+namespace roc {
+namespace core {
+
+StringBuffer::StringBuffer(IArena& arena)
+    : data_(arena) {
+    clear();
+}
+
+bool StringBuffer::is_empty() const {
+    return len() == 0;
+}
+
+size_t StringBuffer::len() const {
+    return data_.size() - 1;
+}
+
+const char* StringBuffer::c_str() const {
+    return data_.data();
+}
+
+void StringBuffer::clear() {
+    if (!data_.resize(1)) {
+        roc_panic("string buffer: allocation failed");
+    }
+    data_[0] = '\0';
+}
+
+ROC_ATTR_NODISCARD bool StringBuffer::assign(const char* str) {
+    if (!str) {
+        roc_panic("string buffer: null pointer");
+    }
+
+    return assign(str, str + strlen(str));
+}
+
+ROC_ATTR_NODISCARD bool StringBuffer::assign(const char* str_begin, const char* str_end) {
```

Comment:
ROC_ATTR_NODISCARD is not needed for definition (in .cpp), only for declaration (in .h)